sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,753 @@
+object.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental {
+
+struct initial_threshold_key {
+  using value_t = property_value<initial_threshold_key>;
+
+  initial_threshold_key(size_t initialThreshold);
+};
+
+struct maximum_size_key {
+  using value_t = property_value<maximum_size_key>;
+
+  initial_threshold_key(size_t maxSize);
+};
+
+struct read_only_key {
+  using value_t = property_value<read_only_key>;
+
+  read_only_key(bool readOnly);
+};
+
+struct zero_init_key {
+  using value_t = property_value<zero_init_key>;
+
+  read_only_key(bool zeroInit);
+};
+
+struct use_existing_memory_key {
+  using value_t = property_value<use_existing_memory_key>;
+
+  use_existing_memory_key(void *ptr, size_t size);
+};
+
+inline constexpr initial_threshold_key::value_t initial_threshold;
+inline constexpr maximum_size_key::value_t maximum_size;
+inline constexpr read_only_key::value_t read_only;
+inline constexpr zero_init_key::value_t zero_init;
+inline constexpr use_existing_memory_key::value_t use_existing_memory;
+
+}  // ext::oneapi::experimental
+----
+
+|===
+|Property|Description
+
+|`initial_threshold`
+|The `initial_threshold` property specifies the initial deallocation threshold
+ value for the memory pool. If this property is not used the default value is
+ zero, and this can be increased after the memory pool is created by calling
+ `memory_pool::set_new_threshold`.
+
+|`maximum_size`
+|The `maximum_size` property specifies the maximum size of the memory pool,
+ after which any allocation will result in an exception. If the value specified
+ is larger than the implementation can support an exception with the
+ `errc::memory_allocation` error code is thrown. If this property is not used
+ the default value is implementation-defined.
+
+|`read_only`
+|The `read_only` property is a performance hint which asserts that all memory
+ allocations from the memory pool will only ever be read from within SYCL kernel
+ functions, this can be used by the SYCL runtime to optimize for performance.
+
+|`zero_init`
+|The `zero_init` property adds the requirement that all memory allocated to the
+ memory as it is allocated from the source to the memory pool will be
+ initialised to zero. Note there is no guarantee that the memory allocation be
+ re-initialized to zero when it is re-allocated from the pool, so users must
+ re-initialize memory to zero if they wish for later allocations to have this
+ behavior.
+
+|`use_existing_memory`
+|The `use_existing_memory` property adds the requirement that the memory pool
+ will use an existing USM memory allocation provided by the user instead of
+ allocating from the SYCL runtime. This property takes a pointer to a valid USM
+ memory allocation of the same allocation kind as the memory pool is initialized
+ with and the size of that memory allocation. Using this property will
+ implicitly set the `maximum_size` and `initial_threshold` property to be that
```

Comment:
This means that all USM allocations will need to be tracked in SYCL/UR (so that we can associate a ptr with the size of the allocation).

Also, why does this mean that the `initial_threshold` property will also be set?

Reply:
Maybe good to also mention that use of a user provided ptr outside using the mem pool during the lifetime of the mem pool results in undef behaviour.